### ISSUE TITLE
Disable Userland proxy by default

### ISSUE THREAD

**Issue by thaJeztah** [docker/docker#14856]
The userland proxy was made optional in https://github.com/docker/docker/pull/12165 (ported to libnetwork in https://github.com/docker/libnetwork/pull/171), 
but still enabled by default because (if I remember correctly) disabling it caused issues on RHEL6 (cf https://github.com/docker/docker/pull/10676).

Now that we will stop supporting RHEL6 with the coming 1.8 release, I think we can change the default to "disabled". We can still keep the `--userland-proxy` option around so that users are able to *enable* it by setting `--userland-proxy=true`. We can remove the proxy altogether in a future release.

Changing the default might help resolving https://github.com/docker/docker/issues/11185 (possibly others)

**Comment by thaJeztah** [comment]
ping @docker/maintainers @mrjana @mavenugo

**Comment by mrjana** [comment]
@thaJeztah SGTM. I will push a PR once we get a few more thumbs up from @docker/maintainers. 

ping @icecrime as well

**Comment by tiborvass** [comment]
I personally think @thaJeztah's proposal makes sense. +1 on changing default to false but keeping the flag just in case we have reports of other weird issues.

**Comment by LK4D4** [comment]
I think we should check if netlink for Hairpin isn't supported with `--false` and die then.

**Comment by tianon** [comment]
:+1:

**Comment by icecrime** [comment]
I'm all for it, I'm just a bit worried as I don't think so many people are giving `--userland-proxy=false` a try today (I have it on my host since 1.7.0 without issues), but that's what the freeze period is for.

**Comment by thaJeztah** [comment]
@icecrime true, but holding off longer won't help there. We've got the flag to give users a way to restore the old behaviour, so even if it results in issues, we can offer users a solution without having to release an update.

**Comment by icecrime** [comment]
@thaJeztah I agree. Do you want to try this for 1.8.0? I'm assuming we'd need a libnetwork PR (to give us `HairpinModeSupported() bool`)?

**Comment by thaJeztah** [comment]
I think it would be good to have this in 1.8.

@mrjana is already in this discussion and offered to create a PR to accommodate the changes (https://github.com/docker/docker/issues/14856#issuecomment-123775368)

I'll add it to the milestone (unless @calavera has objections)

**Comment by calavera** [comment]
sounds great!

**Comment by chenchun** [comment]
I have a question, does libnetwork or docker checks if a random allocated port is currently used by other processes with  `--userland-proxy=false`? I known previously if failing to start userland-proxy process, such as the port is being allocated by other processes on the host, docker will retry to allocate a new port.

**Comment by thaJeztah** [comment]
@mrjana looks like there are some issues reported with `--userland-proxy=false`; https://github.com/docker/docker/issues/15840 (think there was another one, will add if I remember which one)

**Comment by winggundamth** [comment]
After I used --userland-proxy=false then I have this problem 
https://github.com/docker/docker/issues/5618

**Comment by cpuguy83** [comment]
Thanks @winggundamth, we've actually reverted the change that sets this to default because we saw the same thing on our CI.

I'll re-open this now.

**Comment by winggundamth** [comment]
Thanks @cpuguy83. Hope this can fix soon. In our high load production environment with --userland-proxy=true docker-proxy eat 35% of host memory and continue going up. this really critical for us.

Since I'm not a good coder so I'll be a good tester for this case.

**Comment by Soulou** [comment]
I want to highlight that #5618 is pretty critical and seems to affect event the most recent kernels. But I agree on the fact that the userland proxy should be disabled. The question is, what's the best workaround then?

**Comment by thaJeztah** [comment]
For reference; this was reverted for now in https://github.com/docker/docker/pull/16347, more details can be found on that PR as well

**Comment by mrschyte** [comment]
I'm using docker with the `--iptables=false` flag set, because I need to manually set routing with a custom bridge. I'd like to disable the userland proxy, but since the iptables option is set to false, no hairpin nat rules will be generated. Maybe a 3rd option should be available for iptables, that only adds theese rules.

**Comment by oleynikd** [comment]
@mrschyte I'm having exactly the same issue. Have you found some workarounds?

**Comment by mrschyte** [comment]
Denis, I'm currently using docker by setting a fixed ip for the containers.
This can be done using the lxc options before 1.9, and I think 1.9 has a
command line option for it. Another solution would be to listen to docker
events using a script and generate the rules on the fly, but this can be
quite tricky.
2015. nov. 4. 14:24 ezt írta ("Denis" <notifications@github.com>):

**Comment by Silex** [comment]
Disabling the userland proxy seems to have no effect on my nginx containers... is that normal?

``` shell
root@box: # docker version
Client:
 Version:      1.9.1
 API version:  1.21
 Go version:   go1.4.2
 Git commit:   a34a1d5
 Built:        Fri Nov 20 13:12:04 UTC 2015
 OS/Arch:      linux/amd64

Server:
 Version:      1.9.1
 API version:  1.21
 Go version:   go1.4.2
 Git commit:   a34a1d5
 Built:        Fri Nov 20 13:12:04 UTC 2015
 OS/Arch:      linux/amd64
root@box: # docker daemon --userland-proxy=false
root@box: # docker run -it --rm -p 80:80 nginx 
172.17.0.1 - - [28/Jan/2016:13:47:14 +0000] "GET / HTTP/1.1" 304 0 "-" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Ubuntu Chromium/48.0.2564.82 Chrome/48.0.2564.82 Safari/537.36" "-"
```

Should I create a new issue about it?

**Comment by thaJeztah** [comment]
@Silex you need to change the daemon *configuration* file, just running `docker daemon` on the command line like that will only start a second daemon

**Comment by Silex** [comment]
@thaJeztah: I did that first, and as it was not workign I simply stopped the docker service entirely and ran `docker daemon` manually to be sure it was running with the proper parameters. The paste above are commands run in different terminals.

**Comment by Silex** [comment]
Should I open an issue for the fact that testing on the same computer doesn't show 127.0.0.1 but the docker0 ip instead?

**Comment by sergei-solomonov** [comment]
Same problem, change value of --userland-proxy doesn't helped
OS: 15.10 (Wily Werewolf)

```
root: docker version
Client:
 Version:      1.10.2
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   c3959b1
 Built:        Mon Feb 22 21:40:35 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.10.2
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   c3959b1
 Built:        Mon Feb 22 21:40:35 2016
 OS/Arch:      linux/amd64


docker info
Containers: 0
 Running: 0
 Paused: 0
 Stopped: 0
Images: 10
Server Version: 1.10.2
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 17
 Dirperm1 Supported: true
Execution Driver: native-0.2
Logging Driver: json-file
Plugins:
 Volume: local
 Network: host bridge null
Kernel Version: 4.2.0-19-generic
Operating System: Ubuntu 15.10
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 7.71 GiB
Name: xxx
ID: ---
WARNING: No swap limit support
```

**Comment by JonasT** [comment]
Could you try not to break IPv6 port publishing entirely please?

You really should assign someone to take care of IPv6 and make a more coherent concept... the situation seems to be randomly getting a bit better, then *worse*, and then nothing happens for a really long time. It would really be good if you made up your mind about proper IPv6 integration quickly (in a way that is coherent with how the IPv4 setup works, most importantly EXPOSE and publish either working for both or none in similar ways), and then stuck to it for once.

**Comment by twang2218** [comment]
Is there any update on this one? I got the same source IP problem as #15086, which can be solved by `--userland-proxy=false`. I can add the option to some docker hosts, however, I didn't find an easy way to add the option to the Docker for Mac. So, it would be great if `--userland-proxy` is `false` by default.

**Comment by LK4D4** [comment]
@twang2218 default breaks kernel in a very bad way #5618 and #17443

**Comment by thaJeztah** [comment]
@twang2218 I don't think disabling `--userland-proxy` in Docker for Mac will be possible; there's a different version of the proxy running in there to integrate with OS X (IIUC)

**Comment by twang2218** [comment]
@LK4D4 It's sad to hear that. That means I cannot use `--userland-proxy=false` in production. 

Then, how to solve the source IP problem without using `--userland-proxy=false`? It's quite important to have the client IPs for web apps, such as, returning the content based on the clients geolocation, web log analysis, etc.

**Comment by ernestoalejo** [comment]
@twang2218 The feature is tracked here: https://github.com/kubernetes/features/issues/27

**Comment by justincormack** [comment]
That is correct. At some point we would like to look at alternative ways of
doing this in Docker for Mac, but it is not simple.

**Comment by cirocosta** [comment]
Hey, sorry for resurrecting this issue but what is the overall status on this? Is using `userland-proxy` in a recent kernel still a problem or from https://github.com/moby/moby/issues/5618 we can't know yet?

Thx!

**Comment by thaJeztah** [comment]
@cirocosta yes https://github.com/moby/moby/issues/5618 is still an issue, so we don't disable it by default. You may not hit the issue in your particular setup / use, so if you're not hit by that issue you can disable it (YMMV)

**Comment by Xyaren** [comment]
If I understand correctly, disabling the userland-proxy will stop ipv6 reachability out of the box, as of now, there are no ip6tables rules set.

**Comment by agners** [comment]
It seems that #5618 is indeed resolved in newer kernels, so can `user-landproxy=false` considered safe?

**Comment by leo-novu** [comment]
Can anyone shed some light on the status of this issue?

**Comment by etc0de** [comment]
Isn't IPv6-NAT still missing? There were occasional declarations that somebody wants to look at it, but so far it doesn't seem like that was actually done :woman_shrugging:

**Comment by AkihiroSuda** [comment]
@thaJeztah Is there any blocker to move this forward?

**Comment by etc0de** [comment]
I think https://github.com/moby/moby/issues/25407 / https://github.com/moby/moby/pull/41622 might kind of qualify as blockers

**Comment by AkihiroSuda** [comment]
Thanks for the pointer. Can we disable `userland-proxy` by default when `ipv6` is disabled?

**Comment by etc0de** [comment]
I am not aware of any reason why it would be needed for ipv4. Although it sounds like given ipv6 NAT seems to be only weeks away from a merge maybe it would be more consistent to wait until then and just disable it entirely once that is done. Not that it really matters either way though

**Comment by thaJeztah** [comment]
I added this one to the v21.x planning board (https://github.com/moby/moby/projects/11#card-46682701), to be considered for the next release after v20.10

**Comment by ppmathis** [comment]
As I've just dealt with an issue of Docker not preserving the source IP address for incoming UDP packets over a bridge network, I tried disabling the `userland-proxy` myself as often suggested in various issues. I've noticed during testing that while TCP continues to work smoothly, a container running `nfacctd` no longer received any UDP packets on port 5678 despite there being a valid port-mapping with matching iptables rules.

I then stumbled across https://github.com/moby/libnetwork/issues/2423 and I think this is also an issue that would need fixing before userland-proxy can be safely disabled by default? Basically when any incoming UDP packets hit the machine running Docker before its iptables rules are ready, conntrack stores an incorrect state and until that state expires, none of the packets will ever hit the actual container.

As in my case I was dealing with IPFIX which permanently sends UDP packets, this conntrack state never expires by itself. The only workaround I've found was to run `conntrack -D --proto udp` after starting Docker by modifying the systemd service unit (with a short delay), which just kills the state of existing UDP connections. This could obviously be fine-tuned as described in the issue I linked to only terminate states relevant to a port-mapping, but IMHO the Docker daemon should do this on its own.

Just wanted to mention this in case it is not already tracked.

**Comment by thaJeztah** [comment]
@ppmathis thanks! I edited the top description, and linked some tickets related to disabling the user land proxy, so that those can be looked into before switching the default.

**Comment by thaJeztah** [comment]
I also chatted with the Docker Desktop team, and Docker Desktop now sets up networking without relying on the userland-proxy being enabled, so from their side, switching the default won't be an issue; I _think_ they still set up a proxy though, but through other means.

**Comment by ballerburg9005** [comment]
userland proxy is cancer

**Comment by stackunderflow101** [comment]
Hi, Can I ask the current status of this issue? Is this something you are actively working on?

**Comment by thaJeztah** [comment]
For Docker Desktop for Mac and Windows, we're looking at disabling the proxy in the default configuration; on Desktop we have more "control" / "insight" in the exact configuration / setup in which the daemon runs. For Linux, it's a bit more complicated, because there are many setups/scenarios where the daemon runs, which makes it more risky to change the default (as mentioned in this ticket, we tried changing the default, and got reports from users where changing the default caused issues).

Note that if you want it disabled, you can already do this by setting the `"userland-proxy": false` option in `daemon.json` (https://docs.docker.com/engine/reference/commandline/dockerd/#daemon-configuration-file)

A good starting point to get the default changed would be if people could help testing the scenarios linked in the top description (and other issues linked to this ticket), to verify if they're still relevant and/or if changing the default would cause a regression for those.
If we're able to verify that those scenarios are no longer relevant, I think we can consider giving it another go (and change the default to have the user land proxy "disabled" by default).

**Comment by 4ndreasH** [comment]
I tried to switch off the userland-proxy to have access to the external client IP addresses but my docker network is only reachable from internet with userland-proxy: true. The comments here sound like the reason for my problem is probably IPv6.
My local local network is connected to the internet via DS lite, meaning I only have IPv6 for external access. Does "there is no ipv6 NAT" mean: the userland-proxy has to be switched on if IPv6 has to be used?
To use tools like fail2ban inside docker I need to know the clients IP address. How can this be done with IPv6 in this case?
(Sorry if I misunderstood the problem. I'm not a network guy and appreciate any hints where to read more about the userland proxy).

**Comment by unilynx** [comment]
@4ndreasH you need the userland proxy for ipv6 or set up the necessary firewall rules yourself. 

an alternative may be to use host networking (ie no network isolation at all), but that comes with its own tradeoffs.

**Comment by polarathene** [comment]
Potentially relevant issue: https://github.com/moby/moby/issues/44721#issuecomment-1368603067

If disabling becomes the default, I imagine working around the issue when changing to `userland-proxy: true` will be more of a hassle?

There is a lingering postrouting masquerade rule that `userland-proxy: false` adds, and persists when switching to `userland-proxy: true`, but this rule is not there on a fresh boot of the system when starting the docker daemon with `userland-proxy: true`.

**Comment by polarathene** [comment]
> A good starting point to get the default changed would be if people could help testing the scenarios linked in the top description (and other issues linked to this ticket), to verify if they're still relevant and/or if changing the default would cause a regression for those. - (Ref: https://github.com/moby/moby/issues/14856#issuecomment-922723318)

I went over most of the remaining checklist items and commented on each (_except for multi-port and swarm overlay network_). All but the last one below seem resolved now?

---

## Checklist items

### Can close

> - [ ] https://github.com/moby/moby/issues/21860

https://github.com/moby/moby/issues/21860#issuecomment-1534037615

> _Unable to reproduce. Two examples below show connectivity is working fine either way._
> 
> _**Issue has likely been resolved**. Should be fine to close since no one else has chimed in here or reacted to the reported._

---

> - [ ] https://github.com/moby/moby/issues/28589

https://github.com/moby/moby/issues/28589#issuecomment-1534063820

> _I think **this can be closed?**:_
> 
> - _UDP issues have been resolved._
> - _Port publishing should be done if you're going to bind it anyway._
>   
>   - _If you don't need it for IPv6, then be specific and bind only to IPv4 on the host._
>   - _Your host can access a services ports via IP directly if necessary, instead of binding to another IP._

---

> - [ ] [Cannot disable userland-proxy and go with hairpin #37163](https://github.com/moby/moby/issues/37163)

https://github.com/moby/moby/issues/37163#issuecomment-1534078557

> _Looks like **this can be closed?**_
> 
> _Not much engagement in the issue over the years and reading through it seems there was a misunderstanding?_

---

> - [ ] https://github.com/moby/moby/issues/40518

https://github.com/moby/moby/issues/40518#issuecomment-1534127602

> _Reproduced, seems to be effectively the same issue as #38784 (_which goes more into detail to what's causing it_)_

Can technically close as duplicate?

---

### Still an issue

> - [ ] https://github.com/moby/moby/issues/38784

Summarized and verified in https://github.com/moby/moby/issues/38784#issuecomment-1534155356

`userland-proxy: false` prevents containers connecting across separate docker networks on the same host via a published port on the host IP.

## Additional references in this discussion

> I want to highlight that #5618 is pretty critical and seems to affect event the most recent kernels. - (Ref: https://github.com/moby/moby/issues/14856#issuecomment-143045719)

This was a big issue go over, but I summarized it and my investigation: https://github.com/moby/moby/issues/5618#issuecomment-1537076724

That problem was a [motivator for reverting the previous attempt to disable `userland-proxy` by default](https://github.com/moby/moby/pull/16347#issuecomment-140810650). I found several reproduction examples that no longer produce the error (_at least on the modern Vultr VPS environment I ran them on_).

---

> Changing the default might help resolving #11185 (possibly others) - (Ref: https://github.com/moby/moby/issues/14856#issue-96595767)

It doesn't look like this issue needs to be open anymore: https://github.com/moby/moby/issues/11185#issuecomment-1539313597

If it does, it'd be better served with a new issue, just like #5618 would due to the amount of time that has passed and how the history of discussion is filled with noise that doesn't help towards further investigating / resolving.

It's also considered that `userland-proxy: false` (_or the removal of `userland-proxy` in future_) is the desired fix.

---

> `--userland-proxy=false` breaks kernel in a very bad way https://github.com/moby/moby/issues/5618 and https://github.com/moby/moby/issues/17443 - (Ref: https://github.com/moby/moby/issues/14856#issuecomment-238925313)

This was another lengthy and old discussion, and a common pain point users run into. I also summarized the issue:
- https://github.com/moby/moby/issues/15086#issuecomment-1535767303
- https://github.com/moby/moby/issues/17443#issuecomment-1541072515

Majority of problems there were related to IPv6 and `localhost`.
- Since the issue was created, IPv6 support has improved and we now have `ip6tables` (_just need to drop the `experimental` dependency, and potentially enable by default with `default-address-pools` + ULA subnet_).
- `localhost` concerns with client IP may be tied to hobby deployment or dev environments. Both which could probably avoid the issue by approaching the connection differently. Rarely did anyone expand on context, so I'm more inclined that it's related to misconfiguration + lack of docs / advice.

I believe improved IPv6 docs will resolve this. While macOS / Windows Docker Desktop users and anyone using `localhost` that runs into it, could share more context. Chances are docs can resolve that also.

Like the other additional reference links, I think it should be closed with that summary. Let a fresh new issue(s) continue to track remaining concerns, and gather additional context so those can be addressed (_likely via documentation_).

Related to the quoted comment, two comments after it discuss [`userland-proxy: false` not being viable for "Docker for Mac"](https://github.com/moby/moby/issues/14856#issuecomment-238933107), and [a user saying that is a problem for their production web deployment](https://github.com/moby/moby/issues/14856#issuecomment-238934536) (_that apparently relied on "Docker for Mac"?_). Perhaps a misunderstanding on their part. Later an update about ["Docker Desktop" notes that networking no longer relies on `userland-proxy` being enabled](https://github.com/moby/moby/issues/14856#issuecomment-764518390) and [another comment regarding defaulting to disabled for "Docker Desktop" while Linux is a bit more complicated to have confidence in](https://github.com/moby/moby/issues/14856#issuecomment-922723318).

---

> I then stumbled across https://github.com/moby/libnetwork/issues/2423 and I think this is also an issue that would need fixing before userland-proxy can be safely disabled by default? Basically when any incoming UDP packets hit the machine running Docker before its iptables rules are ready, conntrack stores an incorrect state and until that state expires, none of the packets will ever hit the actual container. - (Ref: https://github.com/moby/moby/issues/14856#issuecomment-753471116)

UDP was another common issue I have seen reported. I think that was resolved earlier this year with https://github.com/moby/moby/pull/44742 , although I'm not sure.

---

> I'm using docker with the `--iptables=false` flag set, because I need to manually set routing with a custom bridge.
> I'd like to disable the userland proxy, but since the iptables option is set to false, no hairpin nat rules will be generated. 
> Maybe a 3rd option should be available for iptables, that only adds these rules. - (Ref: https://github.com/moby/moby/issues/14856#issuecomment-151602429)

Not sure if this is a concern. I think originally before the optional `userland-proxy` setting arrived, there was earlier attempts for a `--hairpin-nat` option that would add these. But if you've disabled `iptables`, does it still make sense to support adding just the hairpin nat rules?

They could perhaps be documented with the other `iptables` info, although [those docs also advise against disabling `iptables` stating it can break Docker and adding `iptables` rules can still happen even if you disable the setting](https://docs.docker.com/network/iptables/#prevent-docker-from-manipulating-iptables).

---

[A comment queried about IPv6 with `userland-proxy`, on a IPv6 only host](https://github.com/moby/moby/issues/14856#issuecomment-1055853280), but this is lacking additional context. It may have been a lack of `ip6tables` for ULA, or potentially needing NDP setup properly for GUA. They admit to not being too familiar with networking, and when NDP is required, some VPS services implicitly manipulate network config when Docker adds/removes a veth interface, which can reset `proxy_ndp`.

**Comment by snajpa** [comment]
There are deployments where `daemon.json` is out of reach of people running common kernels on somewhat "bigger" iron - we at vpsFree.cz are one such example of cost-sharing operation; a common shared kernel for all instances has been the primary reason why can we share our resources so efficiently - we've begun long before there was any Docker, and now that there is, we're getting hurt by the assumption that every kernel administrator can influence every userspace configuration running on top of it, and even more by assumption that Docker and the "new container world" has the kernel for itself and thus is free to set sysctls and other things, which aren't really appropriate in the original OpenVZ/Linux-VServer-like hardware-sharing environments.

Now with `docker-proxy` problem at hand, we run into situations where we have thousands upon thousands of `docker-proxy` processes starting up cumulatively, especially on events like botnet whole-prefixes-discovering drive-by's. It's been getting our machines to weird states where OOM killer doesn't apply and the sudden pressure and churn coming from faulting in and droppping virtual memory mappings on such events, which in addition to letting us enjoy the vma/mmap_lock optimizations fallout to the fullest, still isn't probably resolved in full, as we still get not so rare crashes even with latest 6.9/6.10 kernels.

In those kernels, due to how we use them - people are running LXC/LXD and other containers which do use netns+netfilter still - I am pretty sure there currently are no problems with netdev refcount leaks as we don't run into those anymore,  people run extensive QA environments on our boxes, which do stress the net subsys out substantially at times.

In this particular case, I think it would be nice if `docker-proxy` was skipped for kernels that can reliably do netns teardown. Or if `docker-proxy` could really be made tiny, but that would constitute a rewrite from Go, I think? With 1.1G VM size, that's really a huge difference from the final 1-4 MB Rss it ends up occupying. If we could have those 1-4 MB in virtual size too, there would be nothing that could throw the kernel heuristics off.

We really don't have many other options, either it's properly resolved in upstream, or we'll have to, I don't know, ban Docker and force Podman onto people? We can't really afford to have downtime twice a week, as has become the norm in last months... that really hurts our community. :crying_cat_face:

**Comment by ningmingxiao** [comment]
how about "docker run --userland-proxy=false/true "  let docker cli choose  whether to use docker-proxy.

**Comment by ballerburg9005** [comment]
![Image](https://github.com/user-attachments/assets/a47f0efd-5df5-4b8e-a153-7575ed7b2f98)

**Comment by AkihiroSuda** [comment]
Another benefit in disabling `userland-proxy`: With RootlessKit v3.0 (to be released soon), its default `builtin` port driver can now propagate src IP when `userland-proxy` is disabled:
- https://github.com/rootless-containers/rootlesskit/pull/565

Disabling the userland-proxy by default will depend on:
- https://github.com/moby/moby/issues/52294

**Comment by thaJeztah** [comment]
I think overall we all want it to be disabled by default; I recall there were still some remaining reasons to keep it; possibly the comment that Albin left on this PR; https://github.com/moby/moby/pull/50652#issuecomment-3167980506

> Capturing a discussion we had with @robmry on Slack…
> 
> To not break the few edge-cases where the userland proxy is still required (e.g. IPv4-to-IPv6 port-mappings, or the reverse, etc…), I'm considering changing the semantic of `--userland-proxy=true` such that we stop using it whenever it's not needed. So, in most cases, the proxy wouldn't be started anymore.
> 
> As this semantical change might cause issues for cases we didn't identify yet, we could add an env var to restore the previous behavior (e.g. always start the userland proxy), and communicate that it'll be dropped after one major release if no one reports edge-cases where it's needed but not properly started.
> 
> This means that #50419 would need a new daemon flag `--ip-dnat=<true|false>` (enabled by default) that could be disabled by UFW users such that they get all their port-mappings proxied.
